[SelectionDAG] Treat CopyFromReg as freezing the value

bjope opened the pull request:
The description of CopyFromReg in ISDOpcodes.h says that the input
valus is defined outside the scope of the current SelectionDAG. I
think that means that we basically can treat it as a FREEZE in the
sense that it can be seen as neither being undef nor poison.

Being able to fold freeze(CopyFromReg) into CopyFromReg seems
useful to avoid regressions if we start to introduce freeze
instruction in DAGCombiner/foldBoolSelectToLogic, e.g. to solve
https://github.com/llvm/llvm-project/issues/84653

Things _not_ dealt with in this patch:
- Depending on calling convention an input argument can be passed
  also on the stack and not in a register. If it is allowed to treat
  an argument received in a register as not being poison, then I think
  we want to treat arguments received on the stack the same way. But
  then we need to attribute load instructions, or add explicit FREEZE
  when lowering formal arguments.
- A common pattern is that there is an AssertZext or AssertSext just
  after CopyFromReg. I think that if we treat CopyFromReg as never
  being poison, then it should be allowed to fold
     (freeze(AssertZext(CopyFromReg))) -> AssertZext(CopyFromReg))

bjope committed:
[SelectionDAG] Treat CopyFromReg as freezing the value (#85932)

The description of CopyFromReg in ISDOpcodes.h says that the input
valus is defined outside the scope of the current SelectionDAG. I
think that means that we basically can treat it as a FREEZE in the
sense that it can be seen as neither being undef nor poison.

Being able to fold freeze(CopyFromReg) into CopyFromReg seems
useful to avoid regressions if we start to introduce freeze
instruction in DAGCombiner/foldBoolSelectToLogic, e.g. to solve
https://github.com/llvm/llvm-project/issues/84653

Things _not_ dealt with in this patch:
- Depending on calling convention an input argument can be passed
  also on the stack and not in a register. If it is allowed to treat
  an argument received in a register as not being poison, then I think
  we want to treat arguments received on the stack the same way. But
  then we need to attribute load instructions, or add explicit FREEZE
  when lowering formal arguments.
- A common pattern is that there is an AssertZext or AssertSext just
  after CopyFromReg. I think that if we treat CopyFromReg as never
  being poison, then it should be allowed to fold
     (freeze(AssertZext(CopyFromReg))) -> AssertZext(CopyFromReg))

arsenm commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -205,6 +205,7 @@ enum NodeType {
   /// CopyFromReg - This node indicates that the input value is a virtual or
   /// physical register that is defined outside of the scope of this
   /// SelectionDAG.  The register is available from the RegisterSDNode object.
+  /// Not that CopyFromReg is considered as also freezing the value.
```
Note?

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5012,6 +5012,10 @@ bool SelectionDAG::isGuaranteedNotToBeUndefOrPoison(SDValue Op,
     return true;
 
+  case ISD::CopyFromReg:
+    // Treat CopyFromReg as freezing the value.
```
Just add to the case above handling FrameIndex and co? Don't need the special comment?

nikic commented on llvm/test/CodeGen/X86/freeze-binary.ll:
```diff
@@ -202,13 +202,27 @@ define <4 x i32> @freeze_add_vec(<4 x i32> %a0) nounwind {
 ; X86-NEXT:    paddd {{\.?LCPI[0-9]+_[0-9]+}}, %xmm0
-; X86-NEXT:    paddd {{\.?LCPI[0-9]+_[0-9]+}}, %xmm0
+; X86-NEXT:    movl %ebp, %esp
+; X86-NEXT:    popl %ebp
```
Regression?

bjope replied:
Seems like it can be avoided by a patch like this (and suddenly there are floats in the constant pool, but the asm instructions no longer diff):
```
diff --git a/llvm/lib/Target/X86/X86ISelLowering.cpp b/llvm/lib/Target/X86/X86ISelLowering.cpp
index 35f756ea5e1d..d7539b2d942d 100644
--- a/llvm/lib/Target/X86/X86ISelLowering.cpp
+++ b/llvm/lib/Target/X86/X86ISelLowering.cpp
@@ -8771,7 +8771,7 @@ X86TargetLowering::LowerBUILD_VECTOR(SDValue Op, SelectionDAG &DAG) const {
   // and blend the FREEZE-UNDEF operands back in.
   // FIXME: is this worthwhile even for a single FREEZE-UNDEF operand?
   if (unsigned NumFrozenUndefElts = FrozenUndefMask.popcount();
-      NumFrozenUndefElts >= 2 && NumFrozenUndefElts < NumElems) {
+      NumFrozenUndefElts >= 1 && NumFrozenUndefElts < NumElems) {
     SmallVector<int, 16> BlendMask(NumElems, -1);
     SmallVector<SDValue, 16> Elts(NumElems, DAG.getUNDEF(OpEltVT));
     for (unsigned i = 0; i < NumElems; ++i) {

```

bjope replied:
I think the more general problem here is that we now fold freeze over BUILD_VECTOR more often.
Right here, with this patch, we get
```
        t2: v4i32,ch = CopyFromReg t0, Register:v4i32 %0
          t18: i32 = freeze undef:i32
        t19: v4i32 = BUILD_VECTOR Constant:i32<1>, Constant:i32<2>, Constant:i32<3>, t18
      t8: v4i32 = add t2, t19

```
instead of
```
        t2: v4i32,ch = CopyFromReg t0, Register:v4i32 %0
          t7: v4i32 = BUILD_VECTOR Constant:i32<1>, Constant:i32<2>, Constant:i32<3>, undef:i32
        t8: v4i32 = add t2, t7
      t9: v4i32 = freeze t8
```

And then I guess there might be several places where we do handle undef arguments to a BUILD_VECTOR, but we do not deal with "isFreezeUndef" operands just as good. For exampe isBuildVectorOfConstantSDNodes would fail when there is a frozen undef operand, while it would return true if some operands are just undef.

Seems really complicated (and lots of work) to fix all regressions that would pop up if merging https://github.com/llvm/llvm-project/pull/84924 , but considering that different users have suffered from miscompiles I think we might need to accept some regressions ("slow but correct" is often better that "fast but wrong" IMHO).

nikic replied:
I think the right place to fix this would be in generic BUILD_VECTOR lowering to constant pool. In that case, we'd want to lower a BUILD_VECTOR that is constant apart from one-use freeze undef to a constant pool load where the freeze undef has been replaced by zero.

A more aggressive alternative would be to just always replace freeze undef with zero. This is essentially what InstCombine does in the middle-end (well, it's a tad smarter and sometimes uses -1 based on usage). But I suspect that just always doing this may not work as well on the SDAG layer.

bjope replied:
The discussion here https://discourse.llvm.org/t/what-are-we-allowed-to-do-with-freeze-undef-in-selectiondag/78175/11 was started when I looked at this.

bjope replied:
This is one hack to get rid of those regressions: https://github.com/llvm/llvm-project/pull/85932/commits/e2419b60bf60817b8e9888f9ab52063d952a256b

Maybe that should go into a separate pull-request, if it isn't considered as a too hacky solution.

nikic reviewed (commented):
I believe this is correct.

nikic commented:
> A common pattern is that there is an AssertZext or AssertSext just
> after CopyFromReg. I think that if we treat CopyFromReg as never
> being poison, then it should be allowed to fold
> (freeze(AssertZext(CopyFromReg))) -> AssertZext(CopyFromReg))

If we say that violating a zeroext/signext attribute is undefined behavior rather than producing a poison value, then yes -- with the caveat that we can't just patch the AssertZExt + CopyFromReg pattern. The AssertZExt could have come from somewhere other than the ABI.

arsenm commented:
> > A common pattern is that there is an AssertZext or AssertSext just
> > after CopyFromReg. I think that if we treat CopyFromReg as never
> > being poison, then it should be allowed to fold
> > (freeze(AssertZext(CopyFromReg))) -> AssertZext(CopyFromReg))
> 
> If we say that violating a zeroext/signext attribute is undefined behavior rather than producing a poison value, then yes -- with the caveat that we can't just match the AssertZExt + CopyFromReg pattern. The AssertZExt could have come from somewhere other than the ABI.

This would kind of render AssertSext/AssertZext useless

dtcxzyw commented:
Ping.

nikic commented:
I'd like a mitigation for the BUILD_VECTOR issue, but otherwise this looks good to me.

bjope committed:
[DAGCombiner] Do not always fold FREEZE over BUILD_VECTOR (#85932)

Avoid turning a BUILD_VECTOR that can be recognized as "all zeros",
"all ones" or "constant" into something that depends on
freeze(undef), as that would destroy those properties.

Instead we replace undef by 0/-1 in such vectors, making it possible
to fold away the freeze. We typically use -1 if the BUILD_VECTOR
would identify as "all ones", and otherwise we use the value 0.

dtcxzyw commented:
Reverse ping. At least you can rebase to resolve conflicts.
I believe this patch will eliminate a ton of `andi gpr, 1` instructions on RISCV :)

nikic reviewed (approved):
LGTM after a rebase.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -15379,6 +15379,26 @@ SDValue DAGCombiner::visitFREEZE(SDNode *N) {
+    MVT VT = N0.getSimpleValueType();
+    MVT EltVT = VT.getVectorElementType();
+    if (llvm::ISD::isBuildVectorAllOnes(N0.getNode()))
+      return DAG.getSplatBuildVector(VT, DL, DAG.getConstant(-1, DL, EltVT));
```
Use `DAG.getAllOnesConstant()`?

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -15379,6 +15379,26 @@ SDValue DAGCombiner::visitFREEZE(SDNode *N) {
+    if (llvm::ISD::isBuildVectorOfConstantSDNodes(N0.getNode())) {
+      SmallVector<SDValue, 8> NewVecC;
+      for (const SDValue &Op : N0->op_values())
+        NewVecC.push_back(Op.isUndef() ? DAG.getConstant(0, DL, EltVT) : Op);
```
Using EltVT doesn't account for BUILD_VECTOR implicit truncation - it needs to be:
```cpp
DAG.getConstant(0, DL, Op.getValueType())
```

RKSimon reviewed (commented):
A couple of minors

RKSimon reviewed (approved):
LGTM

DavidSpickett commented:
I have had to revert this due to a test suite failure when compiling for AArch64 Scalable Vectors (SVE), for both specific vector lengths and length agnostic code.

```
clang: ../llvm/llvm/include/llvm/CodeGen/ValueTypes.h:307: MVT llvm::EVT::getSimpleVT() const: Assertion `isSimple() && "Expected a SimpleValueType!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.  Program arguments: /home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang -DNDEBUG -mcpu=neoverse-512tvb -mllvm -scalable-vectorization=preferred -mllvm -treat-scalable-fixed-error-as-warning=false -O3 -Wno-implicit-int -Wno-int-conversion -Wno-implicit-function-declaration -w -MD -MT SingleSource/Regression/C/gcc-c-torture/execute/CMakeFiles/GCC-C-execute-regstack-1.dir/regstack-1.c.o -MF CMakeFiles/GCC-C-execute-regstack-1.dir/regstack-1.c.o.d -o CMakeFiles/GCC-C-execute-regstack-1.dir/regstack-1.c.o -c /home/tcwg-buildbot/worker/clang-aarch64-sve-vla/test/test-suite/SingleSource/Regression/C/gcc-c-torture/execute/regstack-1.c
1.  <eof> parser at end of file
2.  Code generation
3.  Running pass 'Function Pass Manager' on module '/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/test/test-suite/SingleSource/Regression/C/gcc-c-torture/execute/regstack-1.c'.
4.  Running pass 'AArch64 Instruction Selection' on function '@main'
<...>
#10 0x0000aaaac4fa66e4 (anonymous namespace)::DAGCombiner::combine(llvm::SDNode*) DAGCombiner.cpp:0:0
#11 0x0000aaaac4fa40b0 llvm::SelectionDAG::Combine(llvm::CombineLevel, llvm::AAResults*, llvm::CodeGenOptLevel) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x87140b0)
#12 0x0000aaaac5110e18 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x8880e18)
#13 0x0000aaaac510f84c llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x887f84c)
#14 0x0000aaaac510c8cc llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x887c8cc)
#15 0x0000aaaac3532350 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x6ca2350)
#16 0x0000aaaac3a526e4 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x71c26e4)
#17 0x0000aaaac3a5a48c llvm::FPPassManager::runOnModule(llvm::Module&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x71ca48c)
#18 0x0000aaaac3a53080 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x71c3080)
#19 0x0000aaaac4794410 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x7f04410)
#20 0x0000aaaac47b8a50 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x7f28a50)
```
Reproducer: [sve_repro.tar.gz](https://github.com/llvm/llvm-project/files/15146873/sve_repro.tar.gz)

This should reproduce on any machine as long as you target AArch64 with SVE. Our AArch64 builder without SVE did not have this problem.

dtcxzyw commented:
> I have had to revert this due to a test suite failure when compiling for AArch64 Scalable Vectors (SVE), for both specific vector lengths and length agnostic code.
> 
> ```
> clang: ../llvm/llvm/include/llvm/CodeGen/ValueTypes.h:307: MVT llvm::EVT::getSimpleVT() const: Assertion `isSimple() && "Expected a SimpleValueType!"' failed.
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
> Stack dump:
> 0.  Program arguments: /home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang -DNDEBUG -mcpu=neoverse-512tvb -mllvm -scalable-vectorization=preferred -mllvm -treat-scalable-fixed-error-as-warning=false -O3 -Wno-implicit-int -Wno-int-conversion -Wno-implicit-function-declaration -w -MD -MT SingleSource/Regression/C/gcc-c-torture/execute/CMakeFiles/GCC-C-execute-regstack-1.dir/regstack-1.c.o -MF CMakeFiles/GCC-C-execute-regstack-1.dir/regstack-1.c.o.d -o CMakeFiles/GCC-C-execute-regstack-1.dir/regstack-1.c.o -c /home/tcwg-buildbot/worker/clang-aarch64-sve-vla/test/test-suite/SingleSource/Regression/C/gcc-c-torture/execute/regstack-1.c
> 1.  <eof> parser at end of file
> 2.  Code generation
> 3.  Running pass 'Function Pass Manager' on module '/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/test/test-suite/SingleSource/Regression/C/gcc-c-torture/execute/regstack-1.c'.
> 4.  Running pass 'AArch64 Instruction Selection' on function '@main'
> <...>
> #10 0x0000aaaac4fa66e4 (anonymous namespace)::DAGCombiner::combine(llvm::SDNode*) DAGCombiner.cpp:0:0
> #11 0x0000aaaac4fa40b0 llvm::SelectionDAG::Combine(llvm::CombineLevel, llvm::AAResults*, llvm::CodeGenOptLevel) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x87140b0)
> #12 0x0000aaaac5110e18 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x8880e18)
> #13 0x0000aaaac510f84c llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x887f84c)
> #14 0x0000aaaac510c8cc llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x887c8cc)
> #15 0x0000aaaac3532350 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x6ca2350)
> #16 0x0000aaaac3a526e4 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x71c26e4)
> #17 0x0000aaaac3a5a48c llvm::FPPassManager::runOnModule(llvm::Module&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x71ca48c)
> #18 0x0000aaaac3a53080 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x71c3080)
> #19 0x0000aaaac4794410 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x7f04410)
> #20 0x0000aaaac47b8a50 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/home/tcwg-buildbot/worker/clang-aarch64-sve-vla/stage1.install/bin/clang+0x7f28a50)
> ```
> 
> Reproducer: [sve_repro.tar.gz](https://github.com/llvm/llvm-project/files/15146873/sve_repro.tar.gz)
> 
> This should reproduce on any machine as long as you target AArch64 with SVE. Our AArch64 builder without SVE did not have this problem.

I meet a similar issue: https://github.com/dtcxzyw/llvm-codegen-benchmark/pull/6#issuecomment-2075880327. But it was caused by an irregular type `<5 x i48>`.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -15452,6 +15452,26 @@ SDValue DAGCombiner::visitFREEZE(SDNode *N) {
+  // the future, then this special handling can be removed.
+  if (N0.getOpcode() == ISD::BUILD_VECTOR) {
+    SDLoc DL(N0);
+    MVT VT = N0.getSimpleValueType();
```
I'm guessing these need to be `EVT VT = N0.getValueType();`

DavidSpickett commented:
creduced example:
```
extern void f();
long double a, b, c, d;
void e() {
  if (a || c || d || b)
    f();
}
```

bjope commented:
> I have had to revert this due to a test suite failure when compiling for AArch64 Scalable Vectors (SVE), for both specific vector lengths and length agnostic code.
> 

Thanks for the revert, and sorry for the inconvenience.
I'm working on a fix + re-apply.

bjope commented:
Re-applied in https://github.com/llvm/llvm-project/commit/55c6bda01ef5a166a69b43956775272d9d67bda5
